src/metabase/models/dashboard.clj

```diff
@@ -74,19 +74,41 @@
     (assert-valid-parameters dashboard)
     (collection/check-collection-namespace Dashboard (:collection_id dashboard))))
 
-(defn- post-update [dashboard]
-  ;; find any subscriptions for this dashboard and update the name to match
-  (let [affected (db/query
-                  {:select    [:p.id]
+(defn- post-update
+  "Updates the pulses' names and syncs the PulseCards"
+  [dashboard]
+  (let [dashboard-id (:id dashboard)
+        affected (db/query
+                  {:select    [[:p.id :pulse_id] [:pc.card_id :card_id]]
                    :modifiers [:distinct]
                    :from      [[Dashboard :d]]
                    :join      [[DashboardCard :dc] [:= :dc.dashboard_id :d.id]
                                [PulseCard :pc] [:= :pc.dashboard_card_id :dc.id]
                                [Pulse :p] [:= :p.id :pc.pulse_id]]
-                   :where     [:= :d.id (:id dashboard)]})]
-    (when (seq affected)
-      (db/update-where! Pulse {:id [:in (map :id affected)]}
-                        :name (:name dashboard)))))
+                   :where     [:= :d.id dashboard-id]})]
+    (when-let [pulse-ids (seq (distinct (map :pulse_id affected)))]
+      (let [correct-card-ids (->> (db/query {:select [:dc.card_id]
+                                             :modifiers [:distinct]
+                                             :from [[DashboardCard :dc]]
+                                             :where [:= :dc.dashboard_id dashboard-id]})
+                                  (map :card_id)
+                                  (remove nil?)
+                                  set)
+            stale-card-ids   (->> affected
+                                  (map :card_id)
+                                  (remove nil?)
+                                  set)
+            cards-to-add     (set/difference correct-card-ids stale-card-ids)]
+        (db/transaction
+          (db/update-where! Pulse {:id [:in pulse-ids]}
+            :name (:name dashboard))
+          (doseq [pulse-id pulse-ids]
+            (doseq [new-card-id cards-to-add]
+              (pulse-card/create! {:card_id new-card-id
+                                   :pulse_id pulse-id
+                                   :dashboard_card_id (db/select-one-id DashboardCard
+                                                        :card_id new-card-id
+                                                        :dashboard_id dashboard-id)}))))))))
```

Comment:
1.  You can just use a single `(doseq [pulse-id pulse-ids, new-card-id cards-to-add] ...)` form here instead of two nested forms

2.  We should really optimize this code... doing multiple DB calls inside a loop like this doesn't scale well. We can definitely insert multiple rows at a time -- I'd suggest tweaking `pulse-card/create!` to allow creating multiple PulseCards at once with a single `db/insert!` call -- shouldn't be too hard.  For avoiding all the calls to `select-one-id`, I would just fetch all the IDs at once as a map of Card ID -> DashboardCardID. Something like this should do it:

    ```clj
    (let [card-id->dashcard-id (db/select-field->id :card_id DashboardCard :dashboard_id dashboard-id, :card-id [:in cards-to-add])]
      ...)
    ```